### ISSUE TITLE
password_authenticator: hashpw is slow

### ISSUE THREAD

**Issue by duarten** [scylladb/scylla#2887]
*Installation details*
Scylla version (or git commit hash): 2.0+, master

6ms stalls (no stalls above 10ms) reported by stall detector (configured with 5ms for a task_quota of 2ms).

`hashpw` seems very slow.

```c++
hashpw(basic_sstring<char, unsigned int, 15u> const&, basic_sstring<char, unsigned int, 15u> const&) at password_authenticator.cc:?
future<shared_ptr<auth::authenticated_user> > futurize<future<shared_ptr<auth::authenticated_user> > >::apply<auth::password_authenticator::authenticate(std::unordered_map<basic_sstring<char, unsigned int, 15u>, basic_sstring<char, unsigned int, 15u>, std::hash<basic_sstring<char, unsigned int, 15u> >, std::equal_to<basic_sstring<char, unsigned int, 15u> >, std::allocator<std::pair<basic_sstring<char, unsigned int, 15u> const, basic_sstring<char, unsigned int, 15u> > > > const&) const::{lambda(future<shared_ptr<cql3::untyped_result_set> >)#2}, future<shared_ptr<cql3::untyped_result_set> > >(auth::password_authenticator::authenticate(std::unordered_map<basic_sstring<char, unsigned int, 15u>, basic_sstring<char, unsigned int, 15u>, std::hash<basic_sstring<char, unsigned int, 15u> >, std::equal_to<basic_sstring<char, unsigned int, 15u> >, std::allocator<std::pair<basic_sstring<char, unsigned int, 15u> const, basic_sstring<char, unsigned int, 15u> > > > const&) const::{lambda(future<shared_ptr<cql3::untyped_result_set> >)#2}&&, future<shared_ptr<cql3::untyped_result_set> >&&) at password_authenticator.cc:?
_ZN12continuationIZN6futureIJ10shared_ptrIN4cql318untyped_result_setEEEE12then_wrappedIZNK4auth22password_authenticator12authenticateERKSt13unordered_mapI13basic_sstringIcjLj15EESB_St4hashISB_ESt8equal_toISB_ESaISt4pairIKSB_SB_EEEEUlS5_E0_S0_IJS1_INS7_18authenticated_userEEEEEET0_OT_EUlST_E_JS4_EE3runEv at password_authenticator.cc:?
```

```c++
hashpw(basic_sstring<char, unsigned int, 15u> const&, basic_sstring<char, unsigned int, 15u> const&) at password_authenticator.cc:?
future<shared_ptr<auth::authenticated_user> > futurize<future<shared_ptr<auth::authenticated_user> > >::apply<auth::password_authenticator::authenticate(std::unordered_map<basic_sstring<char, unsigned int, 15u>, basic_sstring<char, unsigned int, 15u>, std::hash<basic_sstring<char, unsigned int, 15u> >, std::equal_to<basic_sstring<char, unsigned int, 15u> >, std::allocator<std::pair<basic_sstring<char, unsigned int, 15u> const, basic_sstring<char, unsigned int, 15u> > > > const&) const::{lambda(future<shared_ptr<cql3::untyped_result_set> >)#2}, future<shared_ptr<cql3::untyped_result_set> > >(auth::password_authenticator::authenticate(std::unordered_map<basic_sstring<char, unsigned int, 15u>, basic_sstring<char, unsigned int, 15u>, std::hash<basic_sstring<char, unsigned int, 15u> >, std::equal_to<basic_sstring<char, unsigned int, 15u> >, std::allocator<std::pair<basic_sstring<char, unsigned int, 15u> const, basic_sstring<char, unsigned int, 15u> > > > const&) const::{lambda(future<shared_ptr<cql3::untyped_result_set> >)#2}&&, future<shared_ptr<cql3::untyped_result_set> >&&) at password_authenticator.cc:?
auth::password_authenticator::authenticate(std::unordered_map<basic_sstring<char, unsigned int, 15u>, basic_sstring<char, unsigned int, 15u>, std::hash<basic_sstring<char, unsigned int, 15u> >, std::equal_to<basic_sstring<char, unsigned int, 15u> >, std::allocator<std::pair<basic_sstring<char, unsigned int, 15u> const, basic_sstring<char, unsigned int, 15u> > > > const&) const at ??:?
auth::password_authenticator::new_sasl_challenge() const::plain_text_password_challenge::get_authenticated_user() const at password_authenticator.cc:?
transport::cql_server::connection::process_auth_response(unsigned short, std::experimental::fundamentals_v1::basic_string_view<signed char, std::char_traits<signed char> >, service::client_state) at /usr/src/debug/scylla-enterprise-2017.1.2.rc2/transport/server.cc:767
transport::cql_server::connection::process_request_one(std::experimental::fundamentals_v1::basic_string_view<signed char, std::char_traits<signed char> >, unsigned char, unsigned short, service::client_state, transport::cql_server::connection::tracing_request_type)::{lambda()#1}::operator()() at server.cc:?
transport::cql_server::connection::process_request_one(std::experimental::fundamentals_v1::basic_string_view<signed char, std::char_traits<signed char> >, unsigned char, unsigned short, service::client_state, transport::cql_server::connection::tracing_request_type) at ??:?
transport::cql_server::connection::process_request()::{lambda(future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> >&&)#1}::operator()(future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> >&&) const::{lambda(semaphore_units<semaphore_default_exception_factory>)#1}::operator()(semaphore_default_exception_factory) const::{lambda(temporary_buffer<char>)#1}::operator()({lambda(semaphore_units<semaphore_default_exception_factory>)#1})::{lambda()#1}::operator()()::{lambda()#1}::operator()() at server.cc:?
future<> futurize<future<> >::apply<transport::cql_server::connection::process_request()::{lambda(future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> >&&)#1}::operator()(future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> >&&) const::{lambda(semaphore_units<semaphore_default_exception_factory>)#1}::operator()(semaphore_default_exception_factory) const::{lambda(temporary_buffer<char>)#1}, {lambda(semaphore_units<semaphore_default_exception_factory>)#1}>(transport::cql_server::connection::process_request()::{lambda(future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> >&&)#1}::operator()(future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> >&&) const::{lambda(semaphore_units<semaphore_default_exception_factory>)#1}::operator()(semaphore_default_exception_factory) const::{lambda(temporary_buffer<char>)#1}&&, std::tuple<{lambda(semaphore_units<semaphore_default_exception_factory>)#1}>&&) at server.cc:?
future<> futurize<future<> >::apply<transport::cql_server::connection::process_request()::{lambda(future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> >&&)#1}::operator()(future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> >&&) const::{lambda(semaphore_units<semaphore_default_exception_factory>)#1}, semaphore_default_exception_factory>(transport::cql_server::connection::process_request()::{lambda(future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> >&&)#1}::operator()(future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> >&&) const::{lambda(semaphore_units<semaphore_default_exception_factory>)#1}&&, std::tuple<semaphore_default_exception_factory>&&) at server.cc:?
future<> futurize<future<> >::apply<transport::cql_server::connection::process_request()::{lambda(future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> >&&)#1}, future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> > >(transport::cql_server::connection::process_request()::{lambda(future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> >&&)#1}&&, future<std::experimental::fundamentals_v1::optional<transport::cql_binary_frame_v3> >&&) at server.cc:?
_ZN12continuationIZN6futureIJNSt12experimental15fundamentals_v18optionalIN9transport19cql_binary_frame_v3EEEEE12then_wrappedIZNS4_10cql_server10connection15process_requestEvEUlOS7_E_S0_IJEEEET0_OT_EUlSG_E_JS6_EE3runEv at server.cc:?
```

**Comment by slivne** [comment]
@vladzcloudius / @elcallio  / @hakuch - any input on this why was duarte hitting this - is this computed only on connection setup or on every request ? 

- if its only on connection request - then the priority of this should be low 
- if its on every request -is it really needed - I think previously we considered holding a cache of authorized prepared statements per user - @vladzcloudius - not sure if this was done already or not

**Comment by vladzcloudius** [comment]
@slivne We haven't implemented the authorized prepared statements caching yet.

**Comment by hakuch** [comment]
From what I understand, `hashpw` will be called in two main places:

- During any request to authenticate a user, via `auth::check()` (I don't think these are cached, as the permissions cache concerns itself with authorization).
- During requests to `ALTER` and `CREATE` users.

As for the function itself, there are a couple of things:

- Each call allocates a new instance of `crypt_data`, and, quoting from a comment, this is "huge".
- Each call to `gensalt()`, which is used to produce salt for hashing new passwords, allocates a `std::random_device` on the stack.

**Comment by slivne** [comment]
@vladzcloudius / @hakuch  - just so it will be clear - do we need to authenticate the user on every CQL request (e.g. on every EXECUTE of a prepared statement) ?

**Comment by vladzcloudius** [comment]
@slivne Right now - this is what's going to happen.

**Comment by slivne** [comment]
@vladzcloudius - ok thanks - can you reference the issue about the caching of prepared statements per user (if we do not have one - lets create it :) ).

**Comment by elcallio** [comment]
Password checking (and thus hashing) happens (and must happen) every time a new connection (session) is created. In fact, trying to "cache" anything there would be a security risk. 

If crypt_data/random gen do account for the measured time, making them thread_local statics (as suggested in the code comment) should alleviate most of the overhead.

**Comment by duarten** [comment]
Looking at the code, we only authenticate once per connection: `cql_server::connection::make_autheticate` (sic) is only called by `cql_server::connection::process_startup`.

**Comment by duarten** [comment]
Indeed, the client for which these latencies were measured was doing a connection per query.

Still worth to make the improvements suggested by the comment for when we are tested in this way.

**Comment by vladzcloudius** [comment]
@duarten I'm afraid we are talking about different authentication here - you the TLS and me the one against `system_auth`. I think the context of this issue is the TLS authentication. 
@slivne The caching of the permissions you were referring was about caching the permission checks for a user against the `system_auth`, which is irrelevant for this issue.

**Comment by duarten** [comment]
@vladzcloudius Does auth against `system_auth` use `hashpw`?

**Comment by vladzcloudius** [comment]
@duarten Nope.

**Comment by elcallio** [comment]
So crypt_r (the glibc hash function) seems indeed to be very slow. Caching the crypt_data structure does cut about 30% of the hashpw time, but we are still talking (quick dirty measurement) ~2-3 ms per hash call on my machine. Even sha256 should be faster than that for a 10-ish long string.

**Comment by elcallio** [comment]
Dropping hashing rounds to minimum (1000) buys ~10% more. Not highly significant.

**Comment by duarten** [comment]
Realistically, connections should always be reused. Optimizing this further would be more the benefit of simple benchmarking.

**Comment by duarten** [comment]
Fixed by 57c5f13166852849a871ab32d241a67fe973fbc5.
